bobtimus/src/loan.rs

```diff
@@ -44,42 +45,520 @@ pub struct LoanOffer {
+    #[error(
+        "The given price {request_price} is not acceptable with current price {current_price}"
+    )]
+    PriceNotAcceptable {
+        request_price: LiquidUsdt,
+        current_price: LiquidUsdt,
+    },
+
+    #[error("The given principal amount {request_principal} is below the configured minimum {min_principal}")]
+    PrincipalBelowMin {
+        request_principal: LiquidUsdt,
+        min_principal: LiquidUsdt,
+    },
+
+    #[error("The given principal amount {request_principal} is above the configured maximum {max_principal}")]
+    PrincipalAboveMax {
+        request_principal: LiquidUsdt,
+        max_principal: LiquidUsdt,
+    },
+
+    #[error("The LTV value {request_ltv} is above the configured maximum {max_ltv}")]
+    LtvAboveMax {
+        request_ltv: Decimal,
+        max_ltv: Decimal,
+    },
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn validate_loan_is_acceptable(
+    request_price: LiquidUsdt,
+    current_price: LiquidUsdt,
+    price_fluctuation_interval: (Decimal, Decimal),
+    request_principal: LiquidUsdt,
+    min_principal: LiquidUsdt,
+    max_principal: LiquidUsdt,
+    request_ltv: Decimal,
+    max_ltv: Decimal,
+) -> Result<Result<(), LoanValidationError>> {
+    let request_price_dec = Decimal::from(request_price.as_satodollar());
+    let current_price_dec = Decimal::from(current_price.as_satodollar());
+
+    // TODO: Evaluate if we want to use an upper and a lower bound.
+    //  We could just restrict by upper bound, because that is what makes it more expensive for the lender
+    //  i.e. if price was 1000 and is 100 now we must ensure to accept only if the current price it not higher than 100 + x%
+    let (lower, upper) = price_fluctuation_interval;
+    let lower_bound = current_price_dec
+        .checked_mul(lower)
+        .context("multiplication error")?;
+    let upper_bound = current_price_dec
+        .checked_mul(upper)
+        .context("multiplication error")?;
+
+    if request_price_dec < lower_bound || request_price_dec > upper_bound {
+        return Ok(Err(LoanValidationError::PriceNotAcceptable {
+            request_price,
+            current_price,
+        }));
+    }
+
+    if request_principal < min_principal {
+        return Ok(Err(LoanValidationError::PrincipalBelowMin {
+            request_principal,
+            min_principal,
+        }));
+    }
+
+    if request_principal > max_principal {
+        return Ok(Err(LoanValidationError::PrincipalAboveMax {
+            request_principal,
+            max_principal,
+        }));
+    }
+
+    if request_ltv > max_ltv {
+        return Ok(Err(LoanValidationError::LtvAboveMax {
+            request_ltv,
+            max_ltv,
+        }));
+    }
+
+    Ok(Ok(()))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::proptest;
+    use rust_decimal::prelude::FromPrimitive;
+    use rust_decimal_macros::dec;
+
+    #[test]
+    fn test_calculate_interest_rate() {
```

Comment:
I honestly cannot follow your tests. 
Given that these calculations are critical can we have some code comments explaining what is happening?

Reply:
I am not exactly sure what the problem is. I think one problem is, that the logic of making the interest rate configurable on top of loan term and collateralization. 

Additionally, is it problematic that I splitted the calculation into small, testable functions?
At the moment there are tests that make sure that each calculation step is sane - but I did not add tests for the complete loan taking (yet). 

I am not sure if it was the best idea to split the calculation into tiny chunks - I can combine more things again and we can write tests that cover bigger pieces of code - would that help or would you just like to have another bigger test on top?

Reply:
Yes, it could be that the confusing part is how the tests are structured. 

I have a hard time forming the code here into logic without grabbing another piece of paper and writing it out which makes it hard for me personally to see if the logic is sound. 

What I would like to see is something like this (independent of this particular test): 

```rust
// given: 

LTV: 80%
exchange rate start: 10,000 
interest_rate: 5%p.a.
loan duration: 30 days
principal amount: 10,000
overcollateralization_percent: 200%
overcollateralization_btc: 3

// when
do_some_math()

// then
collateralization_thresholds: X,
interest: Y, 
...


```